service/history/historybuilder/event_store.go

```diff
@@ -306,6 +306,13 @@ func (b *EventStore) bufferEvent(
 		enumspb.EVENT_TYPE_WORKFLOW_EXECUTION_UPDATE_COMPLETED:
 		return false
 
+	// Unpaused event should not be buffered since it transitions the workflow from paused to running status.
+	case enumspb.EVENT_TYPE_WORKFLOW_EXECUTION_UNPAUSED:
+		return false
+	// A paused workflow event *should be* allowed to be buffered since we want to accept any inflight workflow task completion.
+	case enumspb.EVENT_TYPE_WORKFLOW_EXECUTION_PAUSED:
+		return true
```

Comment:
why not fail existing workflow task, buffer buffer events and then add the paused event (in the same batch).

Reply:
We made a decision to allow in-flight workflow tasks to complete (mainly to support local activities). So this became necessary.